subsys/debug/mipi_stp_decoder.c

```diff
@@ -0,0 +1,791 @@
+ * @retval opcode	Decoded opcode.
+ */
+static inline uint8_t get_nibble(const uint8_t *data, size_t noff)
+{
+	uint8_t ret = data[noff / 2];
+
+	if (noff & 0x1UL) {
+		ret >>= 4;
+	}
+
+	ret &= 0x0F;
+
+	return ret;
+}
+
+static inline void get_nibbles64(const uint8_t *src, size_t src_noff, uint8_t *dst, size_t dst_noff,
+				 size_t nlen)
+{
+	bool src_ba = (src_noff & 0x1UL) == 0;
+	bool dst_ba = (dst_noff & 0x1UL) == 0;
+	uint64_t *src64 = (uint64_t *)&src[src_noff / 2];
+	uint64_t *dst64 = (uint64_t *)&dst[dst_noff / 2];
+
+	if (nlen == 16) {
+		/* dst must be aligned. */
+		if (src_ba) {
+			uint32_t *s32 = (uint32_t *)src64;
+			uint32_t *d32 = (uint32_t *)dst64;
+
+			d32[0] = s32[0];
+			d32[1] = s32[1];
+		} else {
+			uint64_t part_a = src64[0] >> 4;
+			uint64_t part_b = src64[1] << 60;
+
+			dst64[0] = part_a | part_b;
+		}
+		return;
+	}
+
+	uint64_t mask = BIT64_MASK(nlen * 4) << (src_ba ? 0 : 4);
+	uint64_t src_d = src64[0] & mask;
+
+	if (((src_noff ^ dst_noff) & 0x1UL) == 0) {
+		dst64[0] |= src_d;
+	} else if (dst_ba) {
+		dst64[0] |= (src_d >> 4);
+	} else {
+		dst64[0] |= (src_d << 4);
+	}
+}
+
+/* Function performs getting nibbles in less efficient way but does not use unaligned
+ * access which may not be supported by some platforms.
+ */
+static void get_nibbles_unaligned(const uint8_t *src, size_t src_noff, uint8_t *dst,
+				  size_t dst_noff, size_t nlen)
+{
+	for (size_t i = 0; i < nlen; i++) {
+		size_t idx = (src_noff + i) / 2;
+		size_t ni = (src_noff + i) & 0x1;
+		uint8_t n = src[idx] >> (ni ? 4 : 0);
+		size_t d_idx = (dst_noff + i) / 2;
+		size_t dni = (dst_noff + i) & 0x1;
+
+		if (dni == 0) {
+			dst[d_idx] = n;
+		} else {
+			dst[d_idx] |= n << 4;
+		}
+	}
+}
+
+static inline void get_nibbles(const uint8_t *src, size_t src_noff, uint8_t *dst, size_t dst_noff,
+			       size_t nlen)
+{
+	if (!UNALIGNED_ACCESS_SUPPORTED) {
+		get_nibbles_unaligned(src, src_noff, dst, dst_noff, nlen);
+		return;
+	}
+
+	bool src_ba = (src_noff & 0x1UL) == 0;
+	bool dst_ba = (dst_noff & 0x1UL) == 0;
+	uint32_t *src32 = (uint32_t *)&src[src_noff / 2];
+	uint32_t *dst32 = (uint32_t *)&dst[dst_noff / 2];
+
+	if (nlen > 8) {
+		get_nibbles64(src, src_noff, dst, dst_noff, nlen);
+		return;
+	} else if (nlen == 8) {
+		/* dst must be aligned. */
+		if (src_ba) {
+			dst32[0] = src32[0];
+		} else {
+			uint32_t part_a = src32[0] >> 4;
+			uint32_t part_b = src32[1] << 28;
+
+			dst32[0] = part_a | part_b;
+		}
+		return;
+	}
+
+	uint32_t mask = BIT_MASK(nlen * 4) << (src_ba ? 0 : 4);
+	uint32_t src_d = src32[0] & mask;
+
+	if (((src_noff ^ dst_noff) & 0x1UL) == 0) {
+		dst32[0] |= src_d;
+	} else if (dst_ba) {
+		dst32[0] |= (src_d >> 4);
+	} else {
+		dst32[0] |= (src_d << 4);
+	}
+}
+
+/* Function swaps nibbles in a byte. */
+static uint8_t swap8(uint8_t byte)
+{
+	return (byte << 4) | (byte >> 4);
+}
+
+/* Function swaps nibbles in a 16 bit variable. */
+static uint16_t swap16(uint16_t halfword)
```

Comment:
I'd recommend to change names of these functions to for example swap_nibbles8/16/32 to reflect functionality in direct way